Print error if USER is not in the docker group

AndrejOrsula opened the pull request:
Related to https://github.com/osrf/icra2023_ros2_gz_tutorial/pull/17#issuecomment-1552268544, this PR makes all `docker ...` calls in scripts run as `sudo docker ...` if the current `${USER}` is not detected in the `docker` group. Hopefully, this can save some time for participants that install Docker on-site by not having to deal with permission errors if the step was skipped during installation, or by needing to relog/reboot after they fix it (or someone helps them).

The detection is based on https://stackoverflow.com/a/39920850

I tested it by removing myself from the `docker` group (`gpasswd -d ${USER} docker`) and running `docker/run.bash`. As expected, the script fails when run with the original script (error: `docker: Got permission denied while trying to connect to the Docker daemon socket ...`). With this PR, I need to enter my password and then the container starts correctly.

What is your opinion about this? Let me know if you prefer some other syntax for it.

---

On an unrelated topic, I should note that the error itself *could* be lost between the output of `xhost`:
```
❯ docker/run.bash icra2023_tutorial
icra2023_tutorial
non-network local connections being added to access control list
docker: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create": dial unix /var/run/docker.sock: connect: permission denied.
See 'docker run --help'.
non-network local connections being removed from access control list
```
Maybe it is worth suppressing the output of `xhost +`/`xhost -` to reduce confusion for participants? It is not necessary, just a thought...

mabelzhang commented:
The downside with putting `sudo` in the script is that, that means every time someone runs the script in a new terminal, they have to put in a password. That's not very user-friendly. Wouldn't it be easier to just add them to the docker group (following link in your comment here https://github.com/osrf/icra2023_ros2_gz_tutorial/pull/17#issuecomment-1553553341
 ) with `sudo usermod -a -G docker $USER`? That way they only need to put in their password once, and all subsequent times will work.

AndrejOrsula commented:
> The downside with putting `sudo` in front of `docker run` in the script is that, that means every time someone runs the script in a new terminal, they have to put in a password. That's not very user-friendly.
> 
> Wouldn't it be easier to just add the user to the docker group (...) with `sudo usermod -a -G docker $USER`?

I totally agree! Adding user to the docker group is definitely the preferred approach.

This PR is meant more as a fail-safe in case some participants skip that step. With this PR, they would avoid getting the permission error mentioned above at the cost of having to write their password in every terminal session.

For participants that already added the user to the docker group with `sudo usermod -a -G docker $USER`, this PR should have no effect because all `docker ...` commands in scripts would be executed without `sudo` (just as before).

On the other hand, it might introduce unnecessary complexity. Therefore, feel free to close this PR if you think it is redundant in the presence of the instructions in README.

clalancette commented:
Personally, I think we should take a slightly different approach here.  That is, keep the detection at the beginning of the script to see if the user is in the `docker` group.  If they aren't, then print an error message and quit.  That means that we don't have the additional complexity and user-unfriendliness of using sudo, but we also get automatic detection if the user isn't setup properly.  @AndrejOrsula what do you think?

mabelzhang commented:
Sorry I'm late to reply

> This PR is meant more as a fail-safe in case some participants skip that step

Ah, I meant that it could just run the `sudo usermod` at the top, as soon as it detects the user isn't in the group. Then it wouldn't need to use `sudo` at the bottom.

But I like Chris's approach as well. In that case, we can just detect it, and then print a message saying, "error user not in docker group whatever blah, please add your user to the group with this command before retrying the script: [sudo usermod -a -G docker $USER]" Something like that.

Thanks!

AndrejOrsula commented:
> Ah, I meant that it could just run the `sudo usermod` at the top, as soon as it detects the user isn't in the group. Then it wouldn't need to use `sudo` at the bottom.

Oh, I understand now. However, the user would still need to relog/reboot for changes to take effect. I updated the PR to follow the other approach.

> ... keep the detection at the beginning of the script to see if the user is in the `docker` group. If they aren't, then print an error message and quit.

I added the error message to `build.bash` and `run.bash` in 6754d7e. I did not include it in `join.bash` because the user would have already experienced the error when attempting to use `run.bash`.

Sample output of the error message:
```log
❯ docker/run.bash icra2023_tutorial
Error: The current user 'andrej' is not in the docker group. Before trying again, please add your $USER to the docker group with the following command, and then log out and log back in for changes to take effect:

        sudo usermod -a -G docker $USER

icra2023_tutorial
non-network local connections being added to access control list
docker: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create": dial unix /var/run/docker.sock: connect: permission denied.
See 'docker run --help'.
non-network local connections being removed from access control list
```

> If they aren't, then print an error message **and quit**.

I did not add `exit 1` after printing the error yet, as the script would fail at `docker build/run ...` anyway. Do you think it should be added? Thanks.

clalancette commented on docker/build.bash:
```diff
@@ -31,6 +31,14 @@ then
 
+# If the user is not in the docker group, provide an informative error message
+if [[ $(grep /etc/group -e "docker") != *"${USER}"* ]]; then
+  >&2 echo -e "Error: The current user '${USER}' is not in the docker group. \
```
So I just tested this out locally, and it appears that this is a bit too simplistic a view of docker.  That is, in my setup, my username is *not* in the docker setup, but this script works fine for me.  That may have to do with the fact that I'm using `podman-docker` rather than regular `docker`, but for most practical purposes everything works the same.

So with all of that said, I actually think we shouldn't put it in at this time, given that different setups may have different requirements.

AndrejOrsula replied:
> this is a bit too simplistic a view of docker.

I see, you are right. 
I have not heard about `podman-docker` before. It seems interesting. Thanks for sharing!

> So with all of that said, I actually think we shouldn't put it in at this time, given that different setups may have different requirements.

That makes sense. Feel free to close this PR.

clalancette replied:
OK, thanks for the work here anyway!  Closing this.
